--- eggdrop-1.8.3/src/mod/megahal.mod/megahal.c.org	2018-09-23 12:27:07.121843921 +0200
+++ eggdrop-1.8.3/src/mod/megahal.mod/megahal.c	2018-09-23 12:27:39.979508041 +0200
@@ -291,13 +291,13 @@ static void strip_mirc_codes(int flags,
       }
       break;
     case 0x16:                 /* Reverse video */
-      if (flags & STRIP_REV) {
+      if (flags & STRIP_REVERSE) {
         text++;
         continue;
       }
       break;
     case 0x1f:                 /* Underlined text */
-      if (flags & STRIP_UNDER) {
+      if (flags & STRIP_UNDERLINE) {
         text++;
         continue;
       }
@@ -336,7 +336,7 @@ static void do_megahal(int idx, char *pr
 
   strcpy(buff, text);
   text = buff;
-  strip_mirc_codes(STRIP_BOLD | STRIP_REV | STRIP_UNDER | STRIP_COLOR, text);
+  strip_mirc_codes(STRIP_BOLD | STRIP_REVERSE | STRIP_UNDERLINE | STRIP_COLOR, text);
 
   putlog(LOG_DEBUG, "*", "MegaHAL: learning: %s", text);
 
