--- eggdrop-1.6.10.orig/scripts/autobotchk
+++ eggdrop-1.6.10/scripts/autobotchk
@@ -135,12 +135,12 @@
   puts " *** ERROR: unable to open the file, permission denied"
   puts ""
   exit
- } elseif {![file exists $dir/help] || ![file isdirectory $dir/help]} {
-  puts " *** ERROR: are you sure you are running from a bot directory?"
+ } elseif {![file exists /usr/share/eggdrop/help] || ![file isdirectory /usr/share/eggdrop/help]} {
+  puts " *** ERROR: can't find /usr/share/eggdrop/help!"
   puts ""
   exit
- } elseif {![file exists $dir/$binary]} {
-  puts " *** ERROR: are you sure you are running from a bot directory?"
+ } elseif {![file exists /usr/bin/eggdrop]} {
+  puts " *** ERROR: can't find /usr/bin/eggdrop!"
   puts ""
   exit
  }
@@ -194,7 +194,7 @@
 botdir=\"$dir\"
 
 # change this to the name of your bot's script in that directory:
-botscript=\"$binary $x\"
+botscript=\"$x\"
 
 # change this to the nickname of your bot (capitalization COUNTS)
 botname=\"${botnet-nick}\"
