diff -urNp -x '*.orig' eggdrop-1.8.3.org/eggdrop.conf eggdrop-1.8.3/eggdrop.conf
--- eggdrop-1.8.3.org/eggdrop.conf	2021-10-09 23:37:18.504761324 +0200
+++ eggdrop-1.8.3/eggdrop.conf	2021-10-09 23:37:18.581427892 +0200
@@ -594,14 +594,19 @@ die "Please make sure you edit your conf
 set mod-path "/usr/lib/eggdrop/modules/"
 
 
-#### BLOWFISH MODULE ####
+#### ENCRYPTION MODULE ####
 
 # IF YOU DON'T READ THIS YOU MAY RENDER YOUR USERFILE USELESS LATER
 # Eggdrop encrypts its userfile, so users can have secure passwords.
 # Please note that when you change your encryption method later (i.e.
 # using other modules like a md5 module), you can't use your current
 # userfile anymore. Eggdrop will not start without an encryption module.
-loadmodule blowfish
+# WARNING! rijndael temporarily disabled, do not use it.
+# # Choose only one of the following:
+#loadmodule blowfish
+#loadmodule idea
+#loadmodule twofish
+#loadmodule rijndael (do not use it, needs fixing)
 
 # Specify whether to use ECB or CBC operation modes. ECB is legacy for
 # the Eggdrop blowfish module and is used by default. CBC is more recent
