--- eggdrop1.6.21/eggdrop.conf~	2012-03-31 04:26:15.000000000 +0300
+++ eggdrop1.6.21/eggdrop.conf	2012-03-31 04:27:17.370702471 +0300
@@ -1,4 +1,4 @@
-#! /path/to/executable/eggdrop
+#! /usr/bin/eggdrop
 # ^- This should contain a fully qualified path to your Eggdrop executable.
 #
 # $Id$
@@ -185,22 +185,22 @@ set sort-users 0
 
 # Specify here where Eggdrop should look for help files. Don't modify this
 # setting unless you know what you're doing!
-set help-path "help/"
+set help-path "/usr/share/eggdrop/help/"
 
 # Specify here where Eggdrop should look for text files. This is used for
 # certain Tcl and DCC commands.
-set text-path "text/"
+set text-path "/usr/share/eggdrop/"
 
 # Set here a place to store temporary files.
 set temp-path "/tmp"
 
 # The MOTD (Message Of The day) is displayed when people dcc chat or telnet
 # to the bot. Look at doc/TEXT-SUBSTITUTIONS for options.
-set motd "text/motd"
+set motd "/usr/share/eggdrop/motd"
 
 # This banner will be displayed on telnet connections. Look at
 # doc/TEXT-SUBSTITUTIONS for options.
-set telnet-banner "text/banner"
+set telnet-banner "/usr/share/eggdrop/banner"
 
 # This specifies what permissions the user, channel, and notes files should
 # be set to. The octal values are the same as for the chmod system command.
@@ -419,7 +419,7 @@ die "Please make sure you edit your conf
 # If you run the bot from the compilation directory, you will want to set
 # this to "". If you use 'make install' (like all good kiddies do ;), this
 # is a fine default. Otherwise, use your head :)
-set mod-path "modules/"
+set mod-path "/usr/lib/eggdrop/modules/"
 
 
 #### BLOWFISH MODULE ####
@@ -1378,24 +1378,24 @@
 # directory. All scripts should be put there, although you can place them where
 # you like as long as you can supply a fully qualified path to them.
 #
-# source scripts/script.tcl
+# source /usr/share/eggdrop/scripts/script.tcl
 
-source scripts/alltools.tcl
-source scripts/action.fix.tcl
+source /usr/share/eggdrop/scripts/alltools.tcl
+source /usr/share/eggdrop/scripts/action.fix.tcl
 
 # This script enhances Eggdrop's built-in dcc '.whois' command to allow all
 # users to '.whois' their own handle.
-source scripts/dccwhois.tcl
+source /usr/share/eggdrop/scripts/dccwhois.tcl
 
 # This script provides many useful informational functions, like setting
 # users' URLs, e-mail address, ICQ numbers, etc. You can modify it to add
 # extra entries.
-source scripts/userinfo.tcl
+source /usr/share/eggdrop/scripts/userinfo.tcl
 loadhelp userinfo.help
 
 # Use this script for Tcl and Eggdrop backwards compatibility.
 # NOTE: This can also cause problems with some newer scripts.
-#source scripts/compat.tcl
+#source /usr/share/eggdrop/scripts/compat.tcl
 
 # A few IRC networks (EFnet and Undernet) have added some simple checks to
 # prevent drones from connecting to the IRC network. While these checks are
--- eggdrop1.6.19/src/eggdrop.h	2008-02-16 22:41:03.000000000 +0100
+++ eggdrop1.6.19.new/src/eggdrop.h	2008-09-22 10:00:48.000000000 +0200
@@ -75,7 +75,7 @@
 
 
 /* Language stuff */
-#define LANGDIR  "./language" /* language file directory                   */
+#define LANGDIR  "/usr/share/eggdrop/language" /* language file directory                   */
 #define BASELANG "english"    /* language which always gets loaded before
                                  all other languages. You do not want to
                                  change this.                              */
--- eggdrop1.6.19/eggdrop.conf	2008-02-16 21:15:30.000000000 +0100
+++ eggdrop1.6.19-new/eggdrop.conf	2008-09-23 19:10:36.435603076 +0200
@@ -1359,11 +1359,11 @@ if {[info exists net-type]} {
   switch -- ${net-type} {
     "0" {
       # EFnet
-      source scripts/quotepong.tcl
+      source /usr/share/eggdrop/scripts/quotepong.tcl
     }
     "2" {
       # Undernet
-      source scripts/quotepass.tcl
+      source /usr/share/eggdrop/scripts/quotepass.tcl
     }
   }
 }
