diff -burNp eggdrop1.6.19/src/tcl.c eggdrop1.6.19-new/src/tcl.c
--- eggdrop1.6.19/src/tcl.c	2008-02-16 22:41:04.000000000 +0100
+++ eggdrop1.6.19-new/src/tcl.c	2008-09-22 16:45:55.460597054 +0200
@@ -282,7 +282,7 @@ static char *tcl_eggstr(ClientData cdata
       if (!obj)
         return NULL;
       len = 0;
-      bytes = Tcl_GetByteArrayFromObj(obj, &len);
+      bytes = Tcl_GetStringFromObj(obj, &len);
       if (!bytes)
         return NULL;
       s = malloc(len + 1);
@@ -339,7 +339,7 @@ static int utf_converter(ClientData cdat
   utftot += sizeof(char *) * objc;
   objc -= 5;
   for (i = 0; i < objc; i++) {
-    byteptr = (char *) Tcl_GetByteArrayFromObj(objv[i], &len);
+    byteptr = (char *) Tcl_GetStringFromObj(objv[i], &len);
     strings[i] = (char *) nmalloc(len + 1);
     utftot += len + 1;
     strncpy(strings[i], byteptr, len);
