diff -burNp eggdrop1.6.19/eggdrop.conf eggdrop1.6.19-new/eggdrop.conf
--- eggdrop1.6.19/eggdrop.conf	2008-02-16 21:15:30.000000000 +0100
+++ eggdrop1.6.19-new/eggdrop.conf	2008-09-23 11:32:44.879761861 +0200
@@ -422,14 +422,19 @@ die "Please make sure you edit your conf
 set mod-path "modules/"
 
 
-#### BLOWFISH MODULE ####
+#### ENCRYPTION MODULE ####
 
 # IF YOU DON'T READ THIS YOU MAY RENDER YOUR USERFILE USELESS LATER
 # Eggdrop encrypts its userfile, so users can have secure passwords.
 # Please note that when you change your encryption method later (i.e.
 # using other modules like a md5 module), you can't use your current
 # userfile anymore. Eggdrop will not start without an encryption module.
-loadmodule blowfish
+# WARNING! rijndael temporarily disabled, do not use it.
+# # Choose only one of the following:
+#loadmodule blowfish
+#loadmodule idea
+#loadmodule twofish
+#loadmodule rijndael (do not use it, needs fixing)
 
 
 #### DNS MODULE ####
