--- doc/Makefile.in	2004-10-18 00:23:49.173141992 +0200
+++ doc/Makefile.in	2004-10-18 00:30:44.751964408 +0200
@@ -14,7 +14,8 @@
 
 DOCS = ABOUT AUTHORS BANS BOTNET BUG-REPORT COMPILE-GUIDE CONTENTS \
 FIRST-SCRIPT KNOWN-PROBLEMS MODULES PARTYLINE PATCH-HOWTO tcl-commands.doc \
-TEXT-SUBSTITUTIONS TRICKS UPDATES1.6 USERS
+TEXT-SUBSTITUTIONS TRICKS UPDATES1.6 USERS \
+LANGUAGE.finnish LANGUAGE.portuguese LANGUAGE.portuguese.br
 
 MAN1 = man1/eggdrop.1
 
@@ -50,6 +51,10 @@
 		echo "Creating 'doc/settings' subdirectory."; \
 		$(top_srcdir)/misc/mkinstalldirs $(DESTDIR)/doc/settings >/dev/null; \
 	fi
+	@if test ! -d $(DESTDIR)/doc/modules; then
+		echo "Creating 'doc/modules' subdirectory."; \
+		$(top_srcdir)/misc/mkinstalldirs $(DESTDIR)/doc/modules >/dev/null; \
+	fi
 	@echo "Copying docs."
 	@for i in $(DOCS); do \
 		$(INSTALL_DATA) $(srcdir)/$$i $(DESTDIR)/doc/; \
@@ -70,3 +75,4 @@
 			$(INSTALL_DATA) $$i $(DESTDIR)/doc/html/; \
 		done; \
 	fi
+	@cp -r $(srcdir)/modules/* $(DESTDIR)/doc/modules
