--- eggdrop1.6.21/doc/Makefile.in~	2012-03-31 04:32:33.000000000 +0300
+++ eggdrop1.6.21/doc/Makefile.in	2012-03-31 04:32:58.536425562 +0300
@@ -14,7 +14,8 @@
 
 DOCS = ABOUT BANS BOTNET BUG-REPORT COMPILE-GUIDE CONTENTS \
 Changes1.6 FIRST-SCRIPT KNOWN-PROBLEMS MODULES PARTYLINE PATCH-HOWTO \
-TEXT-SUBSTITUTIONS TRICKS USERS Versions WEIRD-MESSAGES tcl-commands.doc
+TEXT-SUBSTITUTIONS TRICKS USERS Versions WEIRD-MESSAGES tcl-commands.doc \
+LANGUAGE.finnish LANGUAGE.portuguese LANGUAGE.portuguese.br
 
 MAN1 = man1/eggdrop.1
 
@@ -50,6 +51,10 @@
 		echo "Creating 'doc/settings' subdirectory."; \
 		$(top_srcdir)/misc/mkinstalldirs $(DEST)/doc/settings >/dev/null; \
 	fi
+	@if test ! -d $(DEST)/doc/modules; then \
+		echo "Creating 'doc/modules' subdirectory."; \
+		$(top_srcdir)/misc/mkinstalldirs $(DEST)/doc/modules >/dev/null; \
+	fi
 	@echo "Copying docs."
 	@for i in $(DOCS); do \
 		$(INSTALL_DATA) $(srcdir)/$$i $(DEST)/doc/; \
@@ -70,3 +75,4 @@
 			$(INSTALL_DATA) $$i $(DEST)/doc/html/; \
 		done; \
 	fi
+	@cp -pr $(srcdir)/modules/* $(DEST)/doc/modules
