diff -burN eggdrop1.6.17.orig/acinclude.m4 eggdrop1.6.17/acinclude.m4
--- eggdrop1.6.17.orig/acinclude.m4	2004-10-18 00:15:56.783956064 +0200
+++ eggdrop1.6.17/acinclude.m4	2004-10-18 00:38:40.107699352 +0200
@@ -757,9 +757,9 @@
   if test "$IRIX" = "yes"; then
     AC_MSG_WARN([Skipping library tests because they CONFUSE IRIX.])
   else
-    AC_CHECK_LIB(socket, socket)
-    AC_CHECK_LIB(nsl, connect)
-    AC_CHECK_LIB(dns, gethostbyname)
+    AC_SEARCH_LIBS(socket, socket)
+    AC_SEARCH_LIBS(connect, nsl)
+    AC_SEARCH_LIBS(gethostbyname, [resolv bind dns])
     AC_CHECK_LIB(dl, dlopen)
     AC_CHECK_LIB(m, tan, EGG_MATH_LIB="-lm")
 
