diff -urN eggdrop1.6.15/eggdrop.conf eggdrop1.6.15-config_encryption/eggdrop.conf
--- eggdrop1.6.15/eggdrop.conf	Mon May  5 00:05:32 2003
+++ eggdrop1.6.15-config_encryption/eggdrop.conf	Tue May  6 15:44:42 2003
@@ -1273,15 +1273,18 @@
 #loadmodule seen
 
 
-#### BLOWFISH MODULE ####
+#### ENCRYPTION MODULE ####
 
 # IF YOU DON'T READ THIS YOU MAY RENDER YOUR USERFILE USELESS LATER
 # Eggdrop encrypts its userfile, so users can have secure passwords.
 # Please note that when you change your encryption method later (i.e.
 # using other modules like a md5 module), you can't use your current
 # userfile anymore. Eggdrop will not start without an encryption module.
+# Choose only one of the following
 #loadmodule blowfish
-
+#loadmodule idea
+#loadmodule twofish
+#loadmodule rijndael
 
 #### ASSOC MODULE ####
 
