diff -urNp -x '*.orig' eggdrop-1.8.3.org/acinclude.m4 eggdrop-1.8.3/acinclude.m4
--- eggdrop-1.8.3.org/acinclude.m4	2018-02-04 17:08:11.000000000 +0100
+++ eggdrop-1.8.3/acinclude.m4	2021-10-09 23:37:18.961427408 +0200
@@ -852,9 +852,9 @@ AC_DEFUN([EGG_CHECK_LIBS],
   if test "$IRIX" = yes; then
     AC_MSG_WARN([Skipping library tests because they CONFUSE IRIX.])
   else
-    AC_CHECK_LIB(socket, socket)
-    AC_CHECK_LIB(nsl, connect)
-    AC_CHECK_LIB(dns, gethostbyname)
+    AC_SEARCH_LIBS(socket, socket)
+    AC_SEARCH_LIBS(connect, nsl)
+    AC_SEARCH_LIBS(gethostbyname, [resolv bind dns])
     AC_CHECK_LIB(dl, dlopen)
     AC_CHECK_LIB(m, tan, EGG_MATH_LIB="-lm")
 
