--- eggdrop1.6.21/src/Makefile.in	2012-03-31 21:43:01.365906412 +0300
+++ eggdrop1.6.21/src/Makefile.in	2012-03-31 21:51:20.130394596 +0300
@@ -35,11 +35,11 @@
 	@echo ""
 	@cd .. && $(MAKE)
 
-../$(EGGEXEC): build_msg $(eggdrop_objs) $(XREQS) compile_md5 compatability
+../$(EGGEXEC): build_msg $(eggdrop_objs) $(XREQS) compatability
 	@echo "Linking eggdrop $(EGGBUILD)."
 	@echo ""
 	@touch mod/mod.xlibs
-	$(LD) -o ../$(EGGEXEC) $(eggdrop_objs) $(MODOBJS) $(XLIBS) md5/md5c.o compat/*.o `cat mod/mod.xlibs` $(XSSL_LIBS) $(XSSL_INCLUDE)
+	$(LD) -o ../$(EGGEXEC) $(eggdrop_objs) $(MODOBJS) $(XLIBS) compat/*.o `cat mod/mod.xlibs` $(XSSL_LIBS) $(XSSL_INCLUDE)
 	$(STRIP) ../$(EGGEXEC)
 	@echo ""
 	@echo "Successful compile: $(EGGEXEC)"
--- eggdrop1.6.21/src/mod/idea.mod/idea.c~	2003-07-01 03:36:24.000000000 +0300
+++ eggdrop1.6.21/src/mod/idea.mod/idea.c	2012-03-31 21:52:35.356441426 +0300
@@ -29,6 +29,7 @@
 #define MAKING_ENCRYPTION
 
 #include "src/mod/module.h"
+#ifndef EGG_SSL_EXT
 /*
 #include "../../md5/md5c.c"
 #include "../../md5/md5.h"
@@ -45,6 +46,7 @@
 #undef MD5_Init
 #define MD5_Init ((void (*)(MD5_CTX *))global[277])
 #undef global
+#endif
 static Function *global = NULL;
 /********************************************/
 
